akka-persistence-query/src/main/scala/akka/persistence/query/typed/EventEnvelope.scala

```diff
@@ -107,6 +134,54 @@ object EventEnvelope {
 
   def unapply[Event](arg: EventEnvelope[Event]): Option[(Offset, String, Long, Option[Event], Long)] =
     Some((arg.offset, arg.persistenceId, arg.sequenceNr, arg.eventOption, arg.timestamp))
+
+  private object EventHolder {
+    private val _undefinedDeserialize: Array[Byte] => Option[Any] =
+      _ => throw new IllegalStateException("deserialize should not be called, this is a bug")
+    private def undefinedDeserialized[Event]: Array[Byte] => Option[Event] =
+      _undefinedDeserialize.asInstanceOf[Array[Byte] => Option[Event]]
+
+    def apply[Event](eventOption: Option[Event]): EventHolder[Event] =
+      new EventHolder(eventOption, None, undefinedDeserialized[Event])
+
+    def apply[Event](eventBytes: Array[Byte], deserialize: Array[Byte] => Option[Event]): EventHolder[Event] =
+      new EventHolder(None, Some(eventBytes), deserialize)
+  }
+  private final class EventHolder[Event] private (
+      private var _eventOption: Option[Event],
+      val eventBytes: Option[Array[Byte]],
+      private val deserialize: Array[Byte] => Option[Event]) {
+    private var _isDeserialized = eventBytes.isEmpty // not important to make volatile
+
+    lazy val eventOption: Option[Event] = {
+      eventBytes match {
+        case None => _eventOption
+        case Some(bytes) =>
+          _isDeserialized = true
+          deserialize(bytes)
+      }
+    }
```

Comment:
Maybe fine, but one concern: with this auto-deserialization happening on access, there is little control over where/on what thread it happens, vs something more explicit.

Reply:
I'm thinking it must be compatible with usage of the existing EventEnvelope. How would something more explicit look like? Would it throw from `def eventOption` and instead carry (bytes, serId, manifest) as an alternative payload? Could be fine if we make sure to deserialize that in all places before handing out the envelope to the user?

Reply:
but we wouldn't always know, e.g. if they use `eventsBySlices` directly without Projections or such

Reply:
Let me try this from from the other side. Identify the places where we want to use it and see how it would look like from there.